### ISSUE TITLE
Function considered undefined inside itself

### ISSUE THREAD

**Issue by Bouska** [microsoft/pyright#10743]
I'm using function attributes to add "static behavior" to a function. However, pylance reports the function as undefined inside itself when the function is a method. If the function is not in a class, the function is correctly  identified.

## Environment data

-   Pylance version: 2025.6.2
-   OS and version: Ubuntu 24.04.2 LTS
-   Python version (& distribution if applicable, e.g. Anaconda): Ubuntu 24.04.2 LTS (via Ubuntu's .deb)

## Code Snippet

```python
class Printer:
    
    def print_once():
        if not hasattr(print_once, "printed"):
            print_once.printed = False

        if not print_once.printed:
            print("Hello!")

```

## Expected behavior

<img width="562" height="137" alt="Image" src="https://github.com/user-attachments/assets/f7b7ac05-a386-4cb8-915c-db26161ad545" />

## Actual behavior

<img width="716" height="188" alt="Image" src="https://github.com/user-attachments/assets/1b76d7f1-29ee-4c28-a359-1258f686eb00" />

## Logs

```
2025-07-23 11:53:11.846 [info] (Client) Pylance async client (2025.6.2) started with python extension (2025.10.1)
2025-07-23 11:53:12.626 [info] [Info  - 11:53:12 AM] (3588578) Server root directory: file:///home/pablomg/.vscode/extensions/ms-python.vscode-pylance-2025.6.2/dist
2025-07-23 11:53:12.626 [info] [Info  - 11:53:12 AM] (3588578) Pylance language server 2025.6.2 (pyright version 1.1.402, commit 9496ae00) starting
2025-07-23 11:53:12.661 [info] [Info  - 11:53:12 AM] (3588578) Starting service instance "<default>" for workspace ""
2025-07-23 11:53:12.681 [info] [Info  - 11:53:12 AM] (3588578) Setting environmentName for service "<default>": "3.12.3 (global)"
2025-07-23 11:53:12.681 [info] [Info  - 11:53:12 AM] (3588578) Setting pythonPath for service "<default>": "/bin/python"
2025-07-23 11:53:12.681 [info] [Info  - 11:53:12 AM] (3588578) No include entries specified; assuming /<default workspace root>
2025-07-23 11:53:12.681 [info] [Info  - 11:53:12 AM] (3588578) Auto-excluding **/node_modules
2025-07-23 11:53:12.682 [info] [Info  - 11:53:12 AM] (3588578) Auto-excluding **/__pycache__
2025-07-23 11:53:12.682 [info] [Info  - 11:53:12 AM] (3588578) Auto-excluding **/.*
2025-07-23 11:53:12.705 [info] [Info  - 11:53:12 AM] (3588578) Assuming Python version 3.12.3.final.0
2025-07-23 11:53:12.949 [info] [Error - 11:53:12 AM] (3588578) File or directory "/<default workspace root>" does not exist.
2025-07-23 11:53:12.950 [info] [Info  - 11:53:12 AM] (3588578) No source files found.
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) Setting environmentName for service "<default>": "3.12.3 (global)"
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) Setting pythonPath for service "<default>": "/bin/python"
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) No include entries specified; assuming /<default workspace root>
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) Auto-excluding **/node_modules
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) Auto-excluding **/__pycache__
2025-07-23 11:53:13.006 [info] [Info  - 11:53:13 AM] (3588578) Auto-excluding **/.*
2025-07-23 11:53:13.040 [info] [Info  - 11:53:13 AM] (3588578) Assuming Python version 3.12.3.final.0
2025-07-23 11:53:13.240 [info] [Error - 11:53:13 AM] (3588578) File or directory "/<default workspace root>" does not exist.
2025-07-23 11:53:13.240 [info] [Info  - 11:53:13 AM] (3588578) No source files found.
2025-07-23 11:53:13.507 [info] [Info  - 11:53:13 AM] (3588578) BG: Priority queue background worker(2) root directory: file:///home/pablomg/.vscode/extensions/ms-python.vscode-pylance-2025.6.2/dist
2025-07-23 11:53:13.507 [info] [Info  - 11:53:13 AM] (3588578) BG: Priority queue background worker(2) started
2025-07-23 11:53:27.103 [info] [Info  - 11:53:27 AM] (3588578) [id_0] Logging started
2025-07-23 11:53:27.134 [info] [Error - 11:53:27 AM] (3588578) File or directory "/<default workspace root>" does not exist.
2025-07-23 11:53:27.134 [info] [Info  - 11:53:27 AM] (3588578) No source files found.
```

**Comment by erictraut** [comment]
Pyright's behavior is correct here. It's telling you that you have a bug in your code.

Based on Python's name resolution rules, the name `print_once` has no definition at this location. If you attempt to run this code, you'll see an error like this:

```
NameError: name 'print_once' is not defined. Did you mean: 'self.print_once'?
```

**Comment by Hassan9255** [comment]
can i contribute to this?

**Comment by Bouska** [comment]
Oh, my bad! I only checked the PEP 232 and the output of pylance while writing the code snippet and forgot to run the given snippet 🤦  But now I'm not sure I understand PEP 232 and how it is implemented in CPython as I would have expected my snippet to be valid, but that is not a pyright issue. Closing the issue.
